### DISCUSSION TITLE
Does RaiseCanExecuteChanged need to be called on the UI thread?

### DISCUSSION THREAD

**Discussion by CodingOctocat** [TheCodeTraveler/AsyncAwaitBestPractices#78] [Q&A] [answered]
I found this question at StackOverflow, should the RaiseCanExecuteChanged method be executed on the UI thread instead?

[Why people use CommandManager.InvalidateRequerySuggested() on ICommands? - StackOverflow](https://stackoverflow.com/a/62275903) 

```
public void RaiseCanExecuteChanged()
{
    Application.Current.Dispatcher.Invoke(() => CanExecuteChanged?.Invoke(this, null)); 
}
```

**Comment by TheCodeTraveler** [comment, marked as answer]
Hi @CodingOctocat!

The answer is "It Depends".

## Non UI Frameworks
No, you do not need to marshall `RaiseCanExecuteChanged` to the main thread

## UI Frameworks
If you are using a UI dependent framework, such as WinForms, WPF, Xamarin, etc.,

**AND** 

`ICommand` is bound to a UI control (eg a `Button`), then yes you will need to invoke `RaiseCanExecuteChanged` on the main thread:

### Xamarin Example

```csharp
public IAsyncCommand ButtonCommand { get; } = new AsyncCommand(ExecuteButtonCommand);

static async Task ExecuteButtonCommand()
{
    // Handle Button Command
}

void OnCanExecuteChanged() => Device.BeginInvokeOnMainThread(ButtonCommand.RaiseCanExecuteChanged);
```

**Reply by CodingOctocat** [reply]
Thank you! I read the source code, so the default implementation of AABP has no platform detection and it works for non-UI projects? If my project is wpf, I need to compile a copy myself?

**Reply by TheCodeTraveler** [reply]
You don't need to compile a copy of AsyncAwaitBestPractices; using the NuGet package works just fine.

On WPF, just make sure to use `Application.Current.Dispatcher.Invoke` when you call `RaiseCanExecuteChanged`.

### WPF Example

```csharp
public IAsyncCommand ButtonCommand { get; } = new AsyncCommand(ExecuteButtonCommand);

static async Task ExecuteButtonCommand()
{
    // Handle Button Command
}

void OnCanExecuteChanged() => Application.Current.Dispatcher.Invoke(ButtonCommand.RaiseCanExecuteChanged);
```

**Reply by CodingOctocat** [reply]
If I'm lazy and I don't want to manually call RaiseCanExecuteChanged to refresh the UI state every time/every command, and I want the RaiseCanExecuteChanged method to be executed automatically after Execute, then maybe I need to change the source code? (although I don't actually need to, just curious)

**Reply by TheCodeTraveler** [reply]
Correct, AsyncAwaitBestPractices will not automatically call `RaiseCanExecuteChanged`

**Reply by CodingOctocat** [reply]
Thanks a lot really, I think I understand that part now!
